### ISSUE TITLE
Hygiene affected when using multiple derive

### ISSUE THREAD

**Issue by antoyo** [rust-lang/rust#47346]
Hi.
I have a proc-macro that generates a `macro_rules!`.
When I only do:
```rust
#[derive(MyCustomDerive)]
struct MyStruct {
}
```
I can use the generated macro.

However, if I add another custom derive like in:
```rust
#[derive(Serialize, MyCustomDerive)]
struct MyStruct {
}
```
I cannot use the macro anymore…
I get the error:
```
cannot find macro `my_macro!` in this scope
```
Strangely, if I do:
```rust
#[derive(MyCustomDerive, Serialize)]
struct MyStruct {
}
```
Putting the `serde` custom derive after mine fix the issue.
I don't know if it is related to `serde` or just a bug with multiple custom derives.
Thanks to fix this issue.

**Comment by antoyo** [comment]
@dtolnay Any info about this?

**Comment by dtolnay** [comment]
I was able to reproduce this but don't have an explanation.

```rust
#[macro_use]
extern crate repro;

#[macro_use]
extern crate serde_derive;

#[derive(Serialize, MyCustomDerive)]
struct MyStruct;

my_macro!();

fn main() {}
```

```rust
extern crate proc_macro;
use proc_macro::TokenStream;

#[proc_macro_derive(MyCustomDerive)]
pub fn f(_input: TokenStream) -> TokenStream {
    "
        macro_rules! my_macro {
            () => {}
        }
    ".parse().unwrap()
}
```

**Comment by Enselic** [comment]
Triage: I see no strong indication to believe this is a bug in rustc, it seems more likely to be caused by serde, in which case it should be reported there, so let's close this issue. Of course feel free to reopen if it can be demonstrated that this can be reproduced without involving serde.

**Comment by dtolnay** [comment]
This was a rustc bug. It was fixed in Rust 1.39 by https://github.com/rust-lang/rust/pull/63667.
